### PR TITLE
[GOBBLIN-1880] handle invalid cron schedules

### PR DESCRIPTION
**Pull request by meethngala**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [ ] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-XXX
    - https://issues.apache.org/jira/browse/GOBBLIN-1880

### Description
- [ ] Here are some details about my PR, including screenshots (if applicable):
We want to handle invalid cron schedules to capture corner edge cases

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
updated unit tests

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java [hunk 1]
```diff
@@ -301,15 +305,17 @@ public static boolean isWithinRange(String cronExpression, int maxNumDaysToSched
       if (nextValidTimeAfter == null) {
         log.warn("Calculation issue for next valid time for expression: {}. Will default to true for within range",
             cronExpression);
-        return true;
+        return false;
```
Comment (phet) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java:308]: please characterize why the change from current/prior behavior--was it a faulty presumption?

a code comment would go a long way in communicating the reasoning.

aso, log msg is now out-of-sync w/ behavior

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java [hunk 2]
```diff
@@ -291,6 +291,10 @@ private boolean addSpecHelperMethod(Spec spec) {
    */
   @VisibleForTesting
   public static boolean isWithinRange(String cronExpression, int maxNumDaysToScheduleWithin) {
+    if (cronExpression == null || cronExpression.trim().isEmpty()) {
+      // If the cron expression is empty or null, return true to capture adhoc flows
+      return true;
+    }
```
Comment (phet) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java:297]: at the core, I understand why we don't continue w/ `cronExpression == null`... but is this really happening?  more canonical would be to have callers avoid calling this w/ `null`.  then add a `Preconditions` here that croaks if `null` is passed in.

**Comment by Will-Lo** [comment]
What would the user experience be if they added an invalid Cron? I think we want to ensure that the user receives an errorcode 400 as well as some information as to what the issue is with their cron. Potentially see `flowSpec.addCompilationError` for how this is handled. Alternatively, consider adding this check in `BaseFlowToJobSpecCompiler`, which is extended by other compilers in GaaS, because it already sort of tries to handle this, but is missing your additional checks to ensure that the schedule is actually valid.

```
    if (flowSpecProperties.containsKey(ConfigurationKeys.JOB_SCHEDULE_KEY) && StringUtils.isNotBlank(
        flowSpecProperties.getProperty(ConfigurationKeys.JOB_SCHEDULE_KEY))) {
      try {
        new CronExpression(flowSpecProperties.getProperty(ConfigurationKeys.JOB_SCHEDULE_KEY));
      } catch (Exception e) {
        log.error("invalid cron schedule: {}", flowSpecProperties.getProperty(ConfigurationKeys.JOB_SCHEDULE_KEY), e);
        flowSpec.getCompilationErrors().add(new FlowSpec.CompilationError(0, "invalid cron schedule: " + flowSpecProperties.getProperty(ConfigurationKeys.JOB_SCHEDULE_KEY) + e.getMessage()));
        return null;
      }
```
https://github.com/apache/gobblin/blob/69d7e0fa4df2499934462854514ddc9ddbfe7dc7/gobblin-service/src/main/java/org/apache/gobblin/service/modules/flow/BaseFlowToJobSpecCompiler.java#L185

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java (hunk 1 above)
Comment (meethngala) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java:308]: I have added the comments clarifying the need for change in my latest commit.

**Comment by meethngala** [comment]
> What would the user experience be if they added an invalid Cron? I think we want to ensure that the user receives an errorcode 400 as well as some information as to what the issue is with their cron. Potentially see `flowSpec.addCompilationError` for how this is handled. Alternatively, consider adding this check in `BaseFlowToJobSpecCompiler`, which is extended by other compilers in GaaS, because it already sort of tries to handle this, but is missing your additional checks to ensure that the schedule is actually valid.
> 
> ```
>     if (flowSpecProperties.containsKey(ConfigurationKeys.JOB_SCHEDULE_KEY) && StringUtils.isNotBlank(
>         flowSpecProperties.getProperty(ConfigurationKeys.JOB_SCHEDULE_KEY))) {
>       try {
>         new CronExpression(flowSpecProperties.getProperty(ConfigurationKeys.JOB_SCHEDULE_KEY));
>       } catch (Exception e) {
>         log.error("invalid cron schedule: {}", flowSpecProperties.getProperty(ConfigurationKeys.JOB_SCHEDULE_KEY), e);
>         flowSpec.getCompilationErrors().add(new FlowSpec.CompilationError(0, "invalid cron schedule: " + flowSpecProperties.getProperty(ConfigurationKeys.JOB_SCHEDULE_KEY) + e.getMessage()));
>         return null;
>       }
> ```
> 
> https://github.com/apache/gobblin/blob/69d7e0fa4df2499934462854514ddc9ddbfe7dc7/gobblin-service/src/main/java/org/apache/gobblin/service/modules/flow/BaseFlowToJobSpecCompiler.java#L185

As part of this PR, I mainly want to cover scenarios wherein we try to schedule an outdated cron which now expired. We are defaulting it to pass true and running into NPE while scheduling the FlowSpec. As part 


> What would the user experience be if they added an invalid Cron? I think we want to ensure that the user receives an errorcode 400 as well as some information as to what the issue is with their cron. Potentially see `flowSpec.addCompilationError` for how this is handled. Alternatively, consider adding this check in `BaseFlowToJobSpecCompiler`, which is extended by other compilers in GaaS, because it already sort of tries to handle this, but is missing your additional checks to ensure that the schedule is actually valid.
> 
> ```
>     if (flowSpecProperties.containsKey(ConfigurationKeys.JOB_SCHEDULE_KEY) && StringUtils.isNotBlank(
>         flowSpecProperties.getProperty(ConfigurationKeys.JOB_SCHEDULE_KEY))) {
>       try {
>         new CronExpression(flowSpecProperties.getProperty(ConfigurationKeys.JOB_SCHEDULE_KEY));
>       } catch (Exception e) {
>         log.error("invalid cron schedule: {}", flowSpecProperties.getProperty(ConfigurationKeys.JOB_SCHEDULE_KEY), e);
>         flowSpec.getCompilationErrors().add(new FlowSpec.CompilationError(0, "invalid cron schedule: " + flowSpecProperties.getProperty(ConfigurationKeys.JOB_SCHEDULE_KEY) + e.getMessage()));
>         return null;
>       }
> ```
> 
> https://github.com/apache/gobblin/blob/69d7e0fa4df2499934462854514ddc9ddbfe7dc7/gobblin-service/src/main/java/org/apache/gobblin/service/modules/flow/BaseFlowToJobSpecCompiler.java#L185

As part of this PR, I am trying to capture cases wherein we try to schedule an outdated cron which has expired and then run into NPE when scheduling the FlowSpec. Previously, we were defaulting it to pass true for any cron expression that wouldn't pass the number of days constraint within the range.

Now, as per your above ask, I don't see this to be a normal user behavior wherein the user would pass in a valid cron schedule which is in the past e.g. `0 0 0 ? * 6L 2020`. But in case we want to fail fast and handle such scenarios too... I would create a separate PR for that and limit this PR to handle cases from the scheduler standpoint alone. Let me know if that makes sense!

**Review by Will-Lo — APPROVED** [review]
